libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+    if (accuracy < 0.0 || accuracy >= 1.5) {
+        return MAV_ODID_TIME_ACC_UNKNOWN;
+    }
+
+    static const MAV_ODID_TIME_ACC mavoutput [15] = {
+        MAV_ODID_TIME_ACC_0_1_SECOND,
+        MAV_ODID_TIME_ACC_0_2_SECOND,
+        MAV_ODID_TIME_ACC_0_3_SECOND,
+        MAV_ODID_TIME_ACC_0_4_SECOND,
+        MAV_ODID_TIME_ACC_0_5_SECOND,
+        MAV_ODID_TIME_ACC_0_6_SECOND,
+        MAV_ODID_TIME_ACC_0_7_SECOND,
+        MAV_ODID_TIME_ACC_0_8_SECOND,
+        MAV_ODID_TIME_ACC_0_9_SECOND,
+        MAV_ODID_TIME_ACC_1_0_SECOND,
+        MAV_ODID_TIME_ACC_1_1_SECOND,
+        MAV_ODID_TIME_ACC_1_2_SECOND,
+        MAV_ODID_TIME_ACC_1_3_SECOND,
+        MAV_ODID_TIME_ACC_1_4_SECOND,
+        MAV_ODID_TIME_ACC_1_5_SECOND,
+    };
+
+    for (int8_t i = 1; i <= 15; i++) {
+        if (accuracy <= 0.1 * i) {
+            return mavoutput[i-1];
+        }
+    }
+
+    // Should not reach this
+    return MAV_ODID_TIME_ACC_UNKNOWN;
+}
+
+// make sure value is within limits of remote ID standard
+uint16_t AP_OpenDroneID::create_speed_horizontal(uint16_t speed) const
+{
+    if (speed > ODID_MAX_SPEED_H) { // constraint function can't be used, because out of range value is invalid
+        speed = ODID_INV_SPEED_H;
+    }
+
+    return speed;
+}
+
+// make sure value is within limits of remote ID standard
+int16_t AP_OpenDroneID::create_speed_vertical(int16_t speed) const
+{
+    if (speed > ODID_MAX_SPEED_V) { // constraint function can't be used, because out of range value is invalid
+        speed = ODID_INV_SPEED_V;
+    } else if (speed < ODID_MIN_SPEED_V) {
+        speed = ODID_INV_SPEED_V;
+    }
+
+    return speed;
+}
+
+// make sure value is within limits of remote ID standard
+float AP_OpenDroneID::create_altitude(float altitude) const
+{
+    if (altitude > ODID_MAX_ALT) { // constraint function can't be used, because out of range value is invalid
+        altitude = ODID_INV_ALT;
+    } else if (altitude < ODID_MIN_ALT) {
+        altitude = ODID_INV_ALT;
+    }
+
+    return altitude;
+}
+
+// make sure value is within limits of remote ID standard
+float AP_OpenDroneID::create_location_timestamp(float timestamp) const
+{
+    if (timestamp > ODID_MAX_TIMESTAMP) { // constraint function can't be used, because out of range value is invalid
+        timestamp = ODID_INV_TIMESTAMP;
+    } else if (timestamp < 0) {
+        timestamp = ODID_INV_TIMESTAMP;
+    }
+
+    return timestamp;
+}
+
+// handle a message from the GCS
+void AP_OpenDroneID::handle_msg(mavlink_channel_t chan, const mavlink_message_t &msg)
+{
+    WITH_SEMAPHORE(_sem);
+
+    switch (msg.msgid) {
+    // only accept ARM_STATUS from the transmitter
+    case MAVLINK_MSG_ID_OPEN_DRONE_ID_ARM_STATUS: {
+        if (chan == _chan) {
+            mavlink_msg_open_drone_id_arm_status_decode(&msg, &arm_status);
+            last_arm_status_ms = AP_HAL::millis();
+        }
+        break;
+    }
+    // accept other messages from the GCS
```

Comment:
In theory, a drone ID transmitter can also be a [receiver](https://mavlink.io/en/services/opendroneid.html#routing).

I agree that most systems probably will not be doing that but should it already now be taken into account? I guess the current code approach would confuse the messages from the receiver with the messages from the GCS.

This could also be handled in later PRs.

Reply:
Will be later PRs.